### PR TITLE
xbutil reset does not work

### PR DESCRIPTION
**Pull request by houlz0507**

### CODE REVIEW

File: src/runtime_src/driver/xclng/drm/xocl/subdev/mb_scheduler.c [hunk 1]
```diff
@@ -2825,12 +2823,41 @@ static void destroy_client(struct platform_device *pdev, void **priv)
...
+
+	mutex_lock(&xdev->dev_lock);
+	list_for_each(ptr, &xdev->ctx_list) {
+		entry = list_entry(ptr, struct client_ctx, link);
+		if (!entry->abort)
+			abort_client(pdev, entry);
+	}
+	mutex_unlock(&xdev->dev_lock);
+
+	wake_up_interruptible(&exec->poll_wait_queue);
```
Comment (maxzhen) [inline src/runtime_src/driver/xclng/drm/xocl/subdev/mb_scheduler.c]: Once a client is aborted, what prevents it from coming back again through execbuf/wait?

File: src/runtime_src/driver/xclng/drm/xocl/subdev/mb_scheduler.c [hunk 2]
```diff
@@ -3421,6 +3448,8 @@ static int mb_scheduler_remove(struct platform_device *pdev)
 	}
 	mutex_destroy(&exec->exec_lock);
 
+	abort_all_clients(pdev);
```
Comment (maxzhen) [inline src/runtime_src/driver/xclng/drm/xocl/subdev/mb_scheduler.c]: We aborted all clients, without releasing them. But we freed exec and did platform_set_drvdata(pdev, NULL); So, later when destroy_client() is called, can we still free the client data structure?

File: src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_drv.c [hunk 3]
```diff
@@ -76,15 +76,30 @@ struct xocl_pci_funcs userpf_pci_ops = {
...
-	xocl_subdev_destroy_by_id(xdev, XOCL_SUBDEV_DMA);
+	ret = xocl_subdev_get_devinfo(xdev, XOCL_SUBDEV_MAILBOX, &mbox, NULL);
+	if (ret) {
+		xocl_err(&pdev->dev, "can not get mailbox subdev");
+		return;
+	}
+	xocl_subdev_destroy_all(xdev);
+	mbox.priv_data = &mbox_priv;
+	mbox.data_len = 1;
+	ret = xocl_subdev_create_one(xdev, &mbox);
```
Comment (maxzhen) [inline src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_drv.c]: After we recreate mailbox subdevice, i don't see xocl re-register the callback function for processing incoming message. Does the board still work after hot reset?

File: src/runtime_src/driver/xclng/xrt/user_gem/shim.cpp [hunk 4]
```diff
@@ -786,14 +786,25 @@ int xocl::XOCLShim::xclGetDeviceInfo2(xclDeviceInfo2 *info)
...
     int ret;
+    std::string err;
 
     if (kind == XCL_RESET_FULL)
         ret = ioctl(mMgtHandle, XCLMGMT_IOCHOTRESET);
     else if (kind == XCL_RESET_KERNEL)
         ret = ioctl(mMgtHandle, XCLMGMT_IOCOCLRESET);
-    else if (kind == XCL_USER_RESET)
+    else if (kind == XCL_USER_RESET) {
+        int dev_offline;
```
Comment (maxzhen) [inline src/runtime_src/driver/xclng/xrt/user_gem/shim.cpp]: need to initialize it to true?

File: src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_drv.c [hunk 5]
```diff
@@ -165,7 +188,17 @@ int xocl_hot_reset(struct xocl_dev *xdev, bool force)
 	return ret;
 }
 
-void xocl_mb_connect(struct xocl_dev *xdev)
+static void xocl_reset_work(struct work_struct *work)
+{
+	struct xocl_dev *xdev = container_of(to_delayed_work(work),
+			struct xocl_dev, core.reset_work);
+
+	xocl_drvinst_offline(xdev, true);
```
Comment (maxzhen) [inline src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_drv.c]: this is done in xocl_hot_reset_ioctl() already, why do we set it again here?

Comment (houlz0507) [reply src/runtime_src/driver/xclng/drm/xocl/userpf/xocl_drv.c]: I thought xocl_reset_work could be called somewhere else in the future. E.g. if xocl_reset_work is called from sysfs, the caller does not have to call offline first. And it is not hurt to call offline twice.

**Pull request merged**